In [2]:
import numpy as np
# Load the "autoreload" extension so that code can change
%load_ext autoreload

#always reload modules so that as you change code in src, it gets loaded
%autoreload 2

from src.data import make_dataset

In [4]:
path = "/Users/jbaripatti/machine_learning_proj/data/interim/"
name = 'data_interim.txt'
filepath = path + name

In [5]:
data = np.loadtxt(filepath, delimiter=",")

In [7]:
#Split data using a reusable function
X_train, y_train, X_test, y_test = make_dataset.split_data(data)

# Random Forest Classifier

In [9]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import RandomizedSearchCV

## Performance of a RF classifier without tunning

In [10]:
rnd_clf = RandomForestClassifier(n_jobs=-1, random_state=42)
rnd_clf.fit(X_train, y_train)

y_pred = rnd_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

/anaconda3/envs/finalprojectML/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.9252681351631876


## Regularization hyperparameters

In [11]:
from sklearn.model_selection import RandomizedSearchCV

model = RandomForestClassifier()

params = { 'criterion': ('gini', 'entropy'), "max_depth": np.arange(2, 100, 1),
          "min_samples_split" : np.arange(2, 100, 1),
          "min_samples_leaf" : np.arange(1, 100, 1), "max_leaf_nodes": np.arange(2, 1000, 10) ,
         }
rnd_search_cv = RandomizedSearchCV(model, params, n_iter=10, verbose=2, 
                                   scoring="accuracy" ,n_jobs=-1, cv=4)
rnd_search_cv.fit(X_train, y_train)


Fitting 4 folds for each of 10 candidates, totalling 40 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:   51.5s
[Parallel(n_jobs=-1)]: Done  40 out of  40 | elapsed:   59.3s finished
/anaconda3/envs/finalprojectML/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomizedSearchCV(cv=4, error_score='raise-deprecating',
          estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators='warn', n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
          fit_params=None, iid='warn', n_iter=10, n_jobs=-1,
          param_distributions={'criterion': ('gini', 'entropy'), 'max_depth': array([ 2,  3, ..., 98, 99]), 'min_samples_split': array([ 2,  3, ..., 98, 99]), 'min_samples_leaf': array([ 1,  2, ..., 98, 99]), 'max_leaf_nodes': array([  2,  12, ..., 982, 992])},
          pre_dispatch='2*n_jobs', random_state=None, refit=True,
          return_train_score='warn', scoring='accuracy', verbose=2)

In [12]:
print('---- Best estimator ----', rnd_search_cv.best_estimator_)

---- Best estimator ---- RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=38, max_features='auto', max_leaf_nodes=472,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=2, min_samples_split=53,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)


In [13]:
rnd_search_cv.best_score_

0.9234029466319401

In [14]:
y_pred = rnd_search_cv.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.9230769230769231


In [16]:
rnd_clf.feature_importances_

array([0.09781749, 0.05530152, 0.06529824, 0.03119307, 0.01022476,
       0.04427589, 0.03624549, 0.00931598, 0.01474041, 0.01241107,
       0.01019269, 0.01037276, 0.03015025, 0.01562117, 0.00884658,
       0.06267385, 0.05439834, 0.0201768 , 0.00688626, 0.00928435,
       0.0108275 , 0.00613581, 0.03141906, 0.00925505, 0.0083167 ,
       0.01286483, 0.02918577, 0.01168214, 0.01038437, 0.01148469,
       0.00833262, 0.06673701, 0.00841531, 0.00667371, 0.01023391,
       0.0091588 , 0.01853354, 0.03463501, 0.00771256, 0.00910407,
       0.00750946, 0.00824122, 0.00880265, 0.00712951, 0.00405103,
       0.00895848, 0.00720976, 0.00795813, 0.00716904, 0.0064513 ])

### The RF classifier achieved a 92% of accuracy both in training and test dataset, meaning that the model does not overfit the training set and the hyperparameters tuning was successful.